iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/named_pipe.hpp

```diff
@@ -104,11 +115,47 @@ class NamedPipe : public DesignPattern::Creation<NamedPipe, IpcChannelError>
               const size_t maxMsgSize = MAX_MESSAGE_SIZE,
               const uint64_t maxMsgNumber = MAX_NUMBER_OF_MESSAGES) noexcept;
 
-    static IpcChannelName_t convertName(const IpcChannelName_t& name) noexcept;
+    template <typename Prefix>
+    static IpcChannelName_t convertName(const Prefix& p, const IpcChannelName_t& name) noexcept;
 
   private:
     cxx::optional<SharedMemoryObject> m_sharedMemory;
-    MessageQueue_t* m_messages = nullptr;
+
+    class NamedPipeData
+    {
+      public:
+        NamedPipeData(bool& isInitialized, IpcChannelError& error, const uint64_t maxMsgNumber) noexcept;
+        NamedPipeData(const NamedPipeData&) = delete;
+        NamedPipeData(NamedPipeData&& rhs) = delete;
+        ~NamedPipeData() noexcept;
+
+        NamedPipeData& operator=(const NamedPipeData&) = delete;
+        NamedPipeData& operator=(NamedPipeData&& rhs) = delete;
+
+        Semaphore& sendSemaphore() noexcept;
+        Semaphore& receiveSemaphore() noexcept;
+
+        bool waitForInitialization() const noexcept;
+        bool hasValidState() const noexcept;
+
+        MessageQueue_t messages;
+
+      private:
+        static constexpr uint64_t SEND_SEMAPHORE = 0U;
+        static constexpr uint64_t RECEIVE_SEMAPHORE = 1U;
+
+        static constexpr uint64_t INVALID_DATA = 0xBAADF00DAFFEDEAD;
+        static constexpr uint64_t VALID_DATA = 0xBAD0FF1CEBEEFBEE;
+        static constexpr units::Duration WAIT_FOR_INIT_TIMEOUT = units::Duration::fromSeconds(1);
+        static constexpr units::Duration WAIT_FOR_INIT_SLEEP_TIME = units::Duration::fromMilliseconds(1);
+
+        std::atomic<uint64_t> initializationGuard{INVALID_DATA};
+        using semaphoreMemory_t = uint8_t[sizeof(Semaphore)];
+        alignas(alignof(Semaphore)) semaphoreMemory_t semaphores[2U];
```

Comment:
Sadly no since I require direct access to uninitialized memory and the optional does not provide me this kind of access.
The semaphore shouldn't be moved and the idea here is to have the memory available and do a placement new with the creation pattern to avoid the move which would come with the standard `create`.

Reply:
The list of reason why I don't like this pattern grows. We need to switch to the good old factory method pattern.

Reply:
@elBoberido but even the factory pattern would be somehow ugly for objects that are neither copyable nor movable.

The user has a place in mind where the object should life and then provides the factory with a pointer to that memory and the arguments.
But the memory handling should be abstracted! Didn't we had the idea for an uninitialized typed array which provides you the raw correctly aligned memory?

This could then be also used as a building block for the list, forward list, vector, optional etc.

Reply:
You are right. This has the same problem. I thought about `expected<Error> whatever(optional<Semaphore>& s, ...)` could be added to the semaphore. Or maybe something generic where only an `emplace` method is expected.

The uninitialized array is quite a low building block and shouldn't be overused. After all, you have to take care of calling the dtor by yourself, take care of the alignment and such stuff. We already have a typed container for uninitialized data which takes care of all of that. It's the optional.